presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3FileSystem.java

```diff
@@ -605,6 +612,9 @@ private static String keyFromPath(Path path)
         if (key.endsWith(PATH_SEPARATOR)) {
             key = key.substring(0, key.length() - PATH_SEPARATOR.length());
         }
+        if (key.getBytes(UTF_8).length > MAX_S3_KEY_LENGTH) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Length of S3 key is longer than %d bytes: %s", MAX_S3_KEY_LENGTH, key));
```

Comment:
If we want to keep it this way then another way we can handle this is to stop the retry driver when we get an HTTP 400 from S3 and bubble that up like we do in other places.  Ideally this should be added to all S3 calls as any of them can fail with a long key.